site/content/blog/2020-09-22-sites-tree-modifiers/index.md

```diff
@@ -0,0 +1,125 @@
+
+## What does the Sites Tree Represent?
+
+The Sites Tree essentially removes the duplicates from the History table, but also does much more.
+It is ZAPs internal representation of the app and if it does not accurately reflect the app then ZAP will not be able to attack it effectively.
+Each node in the tree represents a different piece of functionality in the app.
+By default ZAP will create unique nodes in the tree based on the HTTP method and the parameter names.
+
+This means that the following requests will all end up having different tree nodes:
+
+* `GET https://www.example.com?a=b&c=d`
+* `GET https://www.example.com?a=b&d=c`
+* `POST https://www.example.com?a=b&c=d`
+
+And these requests will end up having the same node:
+
+* `GET https://www.example.com?a=b&c=d`
+* `GET https://www.example.com?a=c&c=c`
+* `GET https://www.example.com?a=e&c=f`
+
+This works pretty well in lots of cases - different HTTP methods and parameter names typically mean different actions while the parameter values usually don’t change the outcome.
+
+## Exceptions to the Rule
+
+There are always exceptions.
+
+Take the case where the action to be taken is actually defined in a parameter value:
+
+* `GET https://www.example.com?a=b&action=add`
+* `GET https://www.example.com?a=b&action=delete`
+
+In this case both URLs will end up in the same Sites Tree node, which means that in practice only one of them will be attacked. The other URL will not be attacked and so any vulnerabilities specific to that action will not be found.
+
+In ZAP the solution to this is to define the ‘action’ parameter as a [Structural Parameter](/docs/desktop/start/features/structparams/) ie a parameter whose value actually affects the structure of an app. Once you have done that (and revisited the URLs) then they will appear in 2 different nodes and ZAP will attack them separately.
+
+Conversely we have the case where a URL path element is actually part of the data and not part of the structure of the app.
+For example the following URLs could all represent the same functionality if the second path element (companyX) is actually data:
+
+* `https://www.example.com/app/company1/aaa?ddd=eee` 
+* `https://www.example.com/app/company2/aaa?ddd=fff` 
+* `https://www.example.com/app/company3/aaa?ddd=ggg`
+
+In this case all 3 URLs will end up in unique nodes and ZAP will attack each of them even though it doesn't really need to. This probably is not a big problem if there are just 3 instances, but in most cases this data will be coming from a database so there could be a huge number of such nodes.
+
+The solution to this is to define the relevant path element as [Data Driven Content](/docs/desktop/start/features/ddc/) ie path elements that really contain data rather than represent part of the app structure. Once you have done that (and revisited the URLs) then they will all be represented as just 1 node and ZAP will only attack them once.
```

Comment:
```suggestion
The solution to this is to define the relevant path element as [Data Driven Content](/docs/desktop/start/features/ddc/) i.e.: path elements that really contain data rather than represent part of the app structure. Once you have done that (and revisited the URLs) then they will all be represented as just 1 node and ZAP will only attack them once.
```